# SEMANA 8 - MODULARIZACIÓN

Esta semana, nuestro objetivo estará enfocado en:

-Conocer módulos clave de la librería estándar de Python que sirven para resolver problemas reales: cálculos matemáticos, generación de números aleatorios, manejo de fechas y validación de texto.

-Aprender a escribir código más *legible y profesional* siguiendo buenas prácticas de la comunidad.

## Expresiones regulares (re)
Las expresiones regulares permiten buscar y validar patrones de texto de forma precisa.

Con el módulo re aprenderás a comprobar si un correo electrónico es válido, encontrar números dentro de un texto o reemplazar ciertas palabras.

## Módulo math
El módulo math ofrece funciones matemáticas avanzadas que complementan a los operadores básicos de Python.

Permite calcular raíces cuadradas, potencias, redondeos, valores absolutos, trigonometría, logaritmos, entre otros.

Es útil tanto para cálculos simples como para problemas más técnicos que aparecerán en proyectos futuros.

## Módulo random
Con el módulo random podemos generar números aleatorios o elegir elementos al azar de una lista.

Esto sirve para simulaciones, juegos, pruebas y experimentos.

Ejemplos prácticos: lanzar un dado, mezclar un mazo de cartas, o seleccionar un ganador de una lista de usuarios.

## Módulo datetime
El módulo datetime permite trabajar con fechas y horas: obtener la fecha actual, calcular cuántos días faltan para una fecha específica, medir diferencias entre tiempos y formatear fechas para mostrarlas al usuario.

Es esencial en aplicaciones reales como agendas, reportes de actividad, caducidad de productos o logs de sistemas.

## Buenas prácticas en Python
Python se diseñó para ser legible y fácil de mantener.

En esta sección veremos cómo escribir código más profesional siguiendo la convención PEP8: nombres de variables claros, indentación consistente, uso de comentarios, docstrings en funciones y clases, y organización básica del código.

También se presentará la idea de “pythonic code”: preferir soluciones simples, legibles y elegantes.

-------------

# pip

pip es el *gestor de paquetes de Python*. Sirve para *instalar, actualizar y desinstalar* librerías externas (de terceros) que *no vienen* en la librería estándar.

Por defecto, pip descarga paquetes desde *PyPI* (Python Package Index), el repositorio público más grande de paquetes Python.

-Un “paquete” puede traer *código, datos* y, a veces, *binarios precompilados* (ruedas / wheels).

-Con pip puedes “descargarla” y usarla en tu código.

-Cada paquete tiene *versiones* (1.0.0, 1.1.2, …) y *dependencias* (otras librerías que necesita).

## ¿Cómo funciona en Google Colab?
El entorno de Colab es *temporal*. Si cierras o se reinicia la sesión, *se pierden* los paquetes instalados y debes *instalarlos de nuevo*.

-Puedes instalar en una celda con !pip install paquete (o con magia de IPython %pip install paquete).

-Si estuviéramos en un entorno virtual con Python, es decir, fuera de Colab, entonces el signo de admiración no es necesario. Se usaría pip directamente.

Recomendación: coloca todas *tus instalaciones en la primera celda del notebook* para que sea fácil volver a ejecutarla.

## Comandos esenciales
### Instalación y uso de una librería

In [1]:
# Instalar la librería 'requests' (sirve para hacer peticiones a internet)
!pip install requests

# Importar y usar
import requests
respuesta = requests.get("https://jsonplaceholder.typicode.com/todos/1")
print(respuesta.json())

{'userId': 1, 'id': 1, 'title': 'delectus aut autem', 'completed': False}


El resultado es mostrar un pequeño JSON con datos de ejemplo.

### Instalar una versión específica


In [ ]:
# Instalar la librería numpy en una versión exacta
!pip install numpy==1.26.4

import numpy as np
print(np.__version__)   # Confirmamos la versión instalada

### Actualizar una librería

In [3]:
# Actualizar pandas a la última versión
!pip install --upgrade pandas

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 91.2/91.2 kB 3.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.4/12.4 MB 117.8 MB/s eta 0:00:00
  Attempting uninstall: pandas
    Found existing installation: pandas 2.2.2
    Uninstalling pandas-2.2.2:
      Successfully uninstalled pandas-2.2.2
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires pandas==2.2.2, but you have pandas 2.3.3 which is incompatible.
shap 0.50.0 requires numpy>=2, but you have numpy 1.26.4 which is incompatible.


### Desinstalar una librería

In [4]:
# Desinstalar requests
!pip uninstall -y requests

Found existing installation: requests 2.32.4
Uninstalling requests-2.32.4:
  Successfully uninstalled requests-2.32.4


### Instalar varias librerías a la vez

In [5]:
!pip install numpy pandas matplotlib

### Buenas prácticas en Colab con pip
Confirma con import

Al instalar, prueba inmediatamente.

In [6]:
import requests
print("requests instalado correctamente")

requests instalado correctamente


### En Colab, es temporal
Si cierras Colab, las librerías se borran. Tendrías que volver a instalar.

----------------

## Ejemplos
No habrá ejercicios a realizar o entregar. Solo sigue las instrucciones y ve ejecutando gradualmente los siguientes ejemplos.

## Ejemplo 1
Instala la librería requests y úsala para descargar un JSON desde internet.

El link a usar es: https://jsonplaceholder.typicode.com/todos/1.




In [7]:
# Instalar
!pip install requests

# Importar y usar
import requests
respuesta = requests.get("https://jsonplaceholder.typicode.com/todos/1")
print(respuesta.json())

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.7/64.7 kB 2.9 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires pandas==2.2.2, but you have pandas 2.3.3 which is incompatible.
google-colab 1.0.0 requires requests==2.32.4, but you have requests 2.32.5 which is incompatible.


{'userId': 1, 'id': 1, 'title': 'delectus aut autem', 'completed': False}


## Ejemplo 2
Instala la librería numpy en la versión 1.26.4 y verifica la versión instalada con print(np.__version__).

El output debe mostrar exactamente 1.26.4.

In [1]:
# Instalar versión específica
!pip install numpy==1.26.4

# Importar y verificar
import numpy as np
print("Versión instalada:", np.__version__)

Versión instalada: 1.26.4


## Ejemplo 3
Instala la librería faker y úsala para generar un nombre de usuario y una contraseña falsa.

In [2]:
# Instalar
!pip install faker

# Importar y usar
from faker import Faker
fake = Faker()

print("Usuario:", fake.user_name())
print("Contraseña:", fake.password())

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 23.5 MB/s eta 0:00:00
Usuario: chapmanjonathan
Contraseña: )6NacZVkYo


----------------

# Módulos math, random y datetime

Estos tres módulos de la *librería estándar de Python* nos permiten resolver problemas reales sin instalar nada extra.

Son muy usados en cálculos, simulaciones, generación de datos y manejo de fechas.

## Módulo math
El módulo math es parte de la *librería estándar de Python* (no hace falta instalar nada).

Nos da funciones matemáticas *más avanzadas* que las operaciones básicas + - * /.

Es útil cuando necesitamos:

-Calcular raíces, potencias y logaritmos.

-Hacer redondeos más precisos.

-Trabajar con constantes como π o e.

-Resolver problemas de trigonometría.

Muy útil en programación científica, pero también en problemas cotidianos: redondeos, cálculos de interés, simulaciones.

## Funciones importantes
### Raíz cuadrada

In [3]:
import math

print(math.sqrt(25))  # 5.0

5.0


### **Potencias**

In [4]:
print(math.pow(2, 3))  # 8.0 (2 elevado a 3)

8.0


### Redondeos

In [5]:
print(math.ceil(3.1))   # 4 → redondea hacia arriba
print(math.floor(3.9))  # 3 → redondea hacia abajo

4
3


### Valor absoluto

In [6]:
print(math.fabs(-7))  # 7.0

7.0


### Logaritmos

In [7]:
print(math.log(100))     # Logaritmo natural (base e)
print(math.log10(100))   # Logaritmo en base 10

4.605170185988092
2.0


### Trigonometría

In [8]:
print(math.sin(math.pi/2))  # 1.0 → seno de 90°
print(math.cos(math.pi))    # -1.0 → coseno de 180°

1.0
-1.0


### Constantes

In [9]:
print(math.pi)  # 3.141592...
print(math.e)   # 2.718281...

3.141592653589793
2.718281828459045


## Módulo random
El módulo random forma parte de la librería estándar de Python, y nos permite generar valores aleatorios.

Se usa mucho en juegos, simulaciones, sorteos y pruebas donde no queremos resultados siempre iguales.

En programación, “aleatorio” en realidad significa pseudoaleatorio. Los números se generan siguiendo un algoritmo, pero parecen aleatorios.



## Funciones más comunes
### random.random()
Devuelve un número decimal entre 0.0 y 1.0.

In [10]:
import random

print(random.random())  # Ejemplo: 0.3745

0.9623342302788968


### random.randint(a, b)
Devuelve un número entero entre a y b (incluye ambos extremos).

In [11]:
import random

print(random.randint(1, 6))  # Como lanzar un dado (1 al 6)

4


### random.choice(lista)
Elige un elemento al azar de una lista.

In [12]:
import random

frutas = ["manzana", "pera", "uva", "plátano"]

print("Fruta al azar:", random.choice(frutas))

Fruta al azar: plátano


### random.choices(lista, k=n)
Elige varios elementos al azar (con repetición).

In [13]:
import random

frutas = ["manzana", "pera", "uva", "plátano"]

# Elegir 5 frutas (pueden repetirse)
print(random.choices(frutas, k=5))

# ['pera', 'pera', 'uva', 'plátano', 'pera']

['plátano', 'manzana', 'plátano', 'manzana', 'manzana']


### random.sample(lista, k=n)
Elige varios elementos al azar, sin repetición.

In [14]:
import random

estudiantes = ["Ana", "Luis", "Sofía", "Carlos", "Elena"]

# Escoger 2 ganadores distintos
ganadores = random.sample(estudiantes, 2)
print("Ganadores:", ganadores) # ['Sofía', 'Luis']

Ganadores: ['Carlos', 'Ana']


### random.shuffle(lista)
Mezcla una lista en su lugar (la lista cambia de orden).

In [15]:
import random

cartas = list(range(1, 11))  # cartas del 1 al 10
print("Antes:", cartas)

random.shuffle(cartas)
print("Después:", cartas)

Antes: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Después: [5, 4, 1, 7, 6, 9, 8, 10, 3, 2]


## Ejemplo integral
Veamos un sorteo de estudiantes para responder en clase.

In [16]:
import random

estudiantes = ["Ana", "Luis", "Sofía", "Carlos", "Elena"]

print("Sorteo individual:", random.choice(estudiantes))
print("Sorteo de 2 ganadores (sin repetición):", random.sample(estudiantes, 2))

Sorteo individual: Ana
Sorteo de 2 ganadores (sin repetición): ['Sofía', 'Ana']


----------------------

## Módulo datetime
El módulo datetime forma parte de la *librería estándar de Python* y se usa para trabajar con *fechas y horas*.

Sirve para:

-Saber la fecha y hora actual.

-Calcular diferencias entre fechas (cuántos días faltan).

-Sumar o restar días a una fecha.

-Dar formato legible a una fecha (ejemplo: 20/09/2025).

-Convertir un texto en una fecha (parsear).

Cuando usamos datetime, tenemos varios tipos de objetos:

-*datetime*. Contiene *fecha y hora*.

-*date*. Solo la *fecha* (año, mes, día).

-*timedelta*. Representa una *diferencia de tiempo* (ej. 7 días).

## Obtener la fecha y hora actual

In [17]:
from datetime import datetime, date

print("Fecha y hora actual:", datetime.now())
print("Solo la fecha:", date.today())

Fecha y hora actual: 2025-12-25 14:28:33.311553
Solo la fecha: 2025-12-25


## Crear una fecha específica

In [18]:
mi_cumple = datetime(2025, 5, 20, 10, 30)
print("Mi cumple:", mi_cumple)

Mi cumple: 2025-05-20 10:30:00


## Sumar o restar tiempo

In [19]:
una_semana = timedelta(days=7)
hoy = datetime.now()

print("Hoy:", hoy)
print("Dentro de 7 días:", hoy + una_semana)
print("Hace 7 días:", hoy - una_semana)

NameError: name 'timedelta' is not defined

## Formatear una fecha (strftime)

In [20]:
ahora = datetime.now()
print(ahora.strftime("%d/%m/%Y"))   # 24/09/2025
print(ahora.strftime("%H:%M:%S"))   # 14:35:12

25/12/2025
14:30:34


Los formatos comunes serían:

%d → día

%m → mes

%Y → año completo

%H → hora (24h)

%M → minutos

%S → segundos

----

## Ejercicio 1
Simula un dado de 6 caras.

Genera un número entero aleatorio entre 1 y 6 y muéstralo.

In [21]:
import random

def lanzar_dado():
    resultado = random.randint(1, 6)
    print("Salió:", resultado)

# Ejemplo
lanzar_dado()

Salió: 1


## Ejercicio 2
Dada una lista de nombres, selecciona dos ganadores distintos.

Usa random.sample.

In [22]:
import random

estudiantes = ["Ana", "Luis", "Sofía", "Carlos", "Elena"]

ganadores = random.sample(estudiantes, 2)
print("Ganadores:", ganadores)

Ganadores: ['Carlos', 'Sofía']


## Ejercicio 3
Calcula cuántos días faltan para Navidad (25 de diciembre del año actual).

In [23]:
from datetime import date

hoy = date.today()
navidad = date(hoy.year, 12, 25)

faltan = (navidad - hoy).days
print("Faltan", faltan, "días para Navidad")

Faltan 0 días para Navidad


## Ejercicio 4
Genera 5 números aleatorios entre 1 y 10.

Calcula el promedio y redondea hacia arriba con math.ceil.

In [24]:
import random, math

numeros = [random.randint(1, 10) for _ in range(5)]
print("Números:", numeros)

promedio = sum(numeros) / len(numeros)
print("Promedio exacto:", promedio)
print("Promedio redondeado hacia arriba:", math.ceil(promedio))

Números: [3, 1, 9, 7, 4]
Promedio exacto: 4.8
Promedio redondeado hacia arriba: 5


## Ejercicio 5
Simula un sistema que guarda una tarea con la fecha y hora actuales.

Cada vez que agregues una tarea, debe imprimirse:

[fecha y hora] - tarea.

In [25]:
from datetime import datetime

def registrar_tarea(tarea):
    ahora = datetime.now()
    print(f"[{ahora.strftime('%d/%m/%Y %H:%M:%S')}] - {tarea}")

# Ejemplo
registrar_tarea("Estudiar Python")
registrar_tarea("Revisar ejercicios del bootcamp")

[25/12/2025 14:33:23] - Estudiar Python
[25/12/2025 14:33:23] - Revisar ejercicios del bootcamp


-------------------

# Expresiones regulares (re)

Una *expresión regular (regex)* es un patrón de texto que permite *buscar, validar o reemplazar cadenas* de manera flexible.

Son como un “lenguaje miniatura” dentro de Python para detectar coincidencias en un texto.

Veamos algunos ejemplos donde se usaría:

-Validar si un correo electrónico tiene el formato correcto.

-Buscar todas las fechas dentro de un texto.

-Reemplazar palabras ofensivas por ***.

-Extraer todos los números de un párrafo.

### Antes de comenzar
No es necesario que domines al 100% el proceso de regex, ya que es un tema complejo que no es el enfoque en este bootcamp.

Lo que es importante es que los conozcas, veas cómo se forman y en su caso, utilices herramientas o sitios en los cuales puedes generarlos. Puedes verlas en la sección de recursos.

Si quieres profundizarlos, adelante. Valen mucho la pena y se utilizan en muchos lenguajes de programación.

Pero, regex ya está lleno de ejemplos listos para usar. No hay que memorizar todo. Solo entenderlos bien y si necesitas armar algo mucho más personalizado, aplicar los conceptos.

Finalmente, no hay ejercicios para realizar en este contenido.

Ahora si, exploremos.

## Sintaxis básica de patrones
Algunos de los símbolos más usados en regex:

. → cualquier carácter.

^ → inicio de línea.

$ → fin de línea.

\d → dígito (0–9).

\w → carácter de palabra (letra, número o guion bajo).

\s → espacio en blanco.

"+" → una o más repeticiones.

"*" → cero o más repeticiones.

? → cero o una repetición (opcional).

{n,m} → entre n y m repeticiones.

## El módulo re en Python
Python incluye el módulo estándar re para trabajar con regex. Los métodos principales son:

### re.search(patrón, texto) → busca la primera coincidencia.

In [26]:
import re

texto = "Hoy tengo 2 reuniones y 3 pendientes."

# Buscar el primer número en el texto
resultado = re.search(r"\d+", texto)

if resultado:
    print("Coincidencia encontrada:", resultado.group())
    print("Posición:", resultado.span())
else:
    print("No se encontró nada.")

# Coincidencia encontrada: 2
# Posición: (10, 11)

Coincidencia encontrada: 2
Posición: (10, 11)


### re.match(patrón, texto) → verifica si el texto empieza con el patrón.

In [27]:
import re

texto = "Python es genial."

print(re.match(r"Python", texto))   # Coincide porque empieza con "Python"
print(re.match(r"genial", texto))   # No coincide porque "genial" está al final


# <re.Match object; span=(0, 6), match='Python'>
# None

<re.Match object; span=(0, 6), match='Python'>
None


### re.findall(patrón, texto) → devuelve todas las coincidencias en una lista.

In [28]:
import re

texto = "El pedido incluye 5 manzanas, 3 peras y 12 naranjas."

numeros = re.findall(r"\d+", texto)
print(numeros)  # ['5', '3', '12']

['5', '3', '12']


### re.sub(patrón, reemplazo, texto) → reemplaza lo que coincide con otro texto.

In [29]:
import re

texto = "Este producto es malo y feo."

# Reemplazar palabras ofensivas
limpio = re.sub(r"malo|feo", "***", texto)
print(limpio)

# Este producto es *** y ***.

Este producto es *** y ***.


### re.split(patrón, texto) → divide un texto en partes según el patrón.

In [30]:
import re

texto = "Hola mundo. ¿Cómo estás? Bien, gracias!"

# Dividir usando signos de puntuación como separadores
frases = re.split(r"[.?!]", texto)
print(frases)

# ['Hola mundo', ' ¿Cómo estás', ' Bien, gracias', '']

['Hola mundo', ' ¿Cómo estás', ' Bien, gracias', '']


### Consejos para crear expresiones regulares
Empieza simple
No intentes escribir un regex largo de una sola vez. Empieza probando cosas pequeñas.

In [31]:
# Buscar solo números
re.findall(r"\d+", "Hay 3 gatos y 25 perros")
# ['3', '25']

['3', '25']

*Piensa en el patrón, no en el programa completo*

En lugar de “quiero validar un correo”, piensa:

-Primero: “necesito letras o números” → \w+

-Luego: “después un @” → @

-Luego: “más letras” → \w+

-Al final: “.com o similar” → \.\w+

-Combinas poco a poco y listo.

*Usa raw strings en Python (r"")*

Siempre escribe los patrones como r"..." para que Python no intente interpretar los \.

In [32]:
# Correcto
r"\d+"

# Incorrecto
"\d+"  # Python se confunde con la barra invertida

<>:5: SyntaxWarning: invalid escape sequence '\d'
<>:5: SyntaxWarning: invalid escape sequence '\d'
/tmp/ipython-input-1080279282.py:5: SyntaxWarning: invalid escape sequence '\d'
  "\d+"  # Python se confunde con la barra invertida


'\\d+'

-------

## Ejemplos prácticos
Buscar un número en un texto

In [33]:
import re

texto = "Hoy tengo 2 reuniones y 3 pendientes."
resultado = re.findall(r"\d+", texto)
print(resultado)  # ['2', '3']

['2', '3']


Validar un correo electrónico

In [34]:
import re

correo = "mike@example.com"
patron = r"^[\w\.-]+@[\w\.-]+\.\w+$"

if re.match(patron, correo):
    print("Correo válido")
else:
    print("Correo inválido")

Correo válido


Reemplazar malas palabras

In [35]:
import re

frase = "Este producto es malo y feo."
limpia = re.sub(r"malo|feo", "***", frase)
print(limpia)  # Este producto es *** y ***.

Este producto es *** y ***.


Extraer todas las palabras de un texto

In [36]:
import re

texto = "Python es genial, rápido y divertido."
palabras = re.findall(r"\w+", texto)
print(palabras)  # ['Python', 'es', 'genial', 'rápido', 'y', 'divertido']

['Python', 'es', 'genial', 'rápido', 'y', 'divertido']


Dividir un texto en frases

In [37]:
import re

texto = "Hola mundo. ¿Cómo estás? Bien, gracias!"
frases = re.split(r"[.?!]", texto)
print(frases)  # ['Hola mundo', ' ¿Cómo estás', ' Bien, gracias', '']

['Hola mundo', ' ¿Cómo estás', ' Bien, gracias', '']


----------

# Buenas prácticas

### PEP 8 - Guía de estilo

A continuación, veremos buenas prácticas que puedes utilizar en Python para tener en tu código mucho más fácil de leer y ejecutar.

PEP 8 es el estándar que define cómo debe escribirse el código en Python.

Los puntos más relevantes son:

-*Indentación*: usar 4 espacios por nivel de indentación (nunca tabs, la idea es que no mezcles espacios con tabs). En editores, puedes configurar para que cuando presiones la tecla "tab" se generen 4 espacios.

-*Longitud de línea*: máximo ~79 caracteres por línea si es posible. Dependerá del caso de uso pero se recomienda.

In [1]:
# Mejor: romper en varias líneas
mensaje = (
    "Este es un mensaje largo que se parte en varias líneas para mantener "
    "la legibilidad del código."
)

### Espacios:

-Alrededor de operadores: x = a + b, no x=a+b.

-Después de coma: [1, 2, 3], no [1,2,3].

In [ ]:
# Correcto
x = a + b
numeros = [1, 2, 3]

# No es incorrecto pero puede ser más difícil de leer
x=a+b
numeros = [1,2,3]

-*Nombres de variables y funciones*. usar snake_case → calcular_total().

-*Nombres de clases*. usar CamelCase → class ProductoCarrito:.

-*Constantes*. todo en mayúsculas → PI = 3.14159.

-*Importaciones*. Un import por línea.

In [3]:
# Bien
import math
import random

# No está mal pero es más difícil de leer
import math, random

### Comentario y Docstrings
Comentarios que expliquen el “por qué”.

Evita comentar lo obvio; explica la intención o la decisión.

In [4]:
# Explica por qué un diccionario (O(1) promedio por búsqueda):
usuarios_por_id = {101: "Ana", 202: "Luis"}

-Docstrings (triple comilla) para funciones/clases

Una línea si es simple; varias si necesitas detallar parámetros y retorno.

In [5]:
def dividir(a, b):
    """Devuelve a/b. Lanza ValueError si b == 0."""
    if b == 0:
        raise ValueError("No se puede dividir entre cero.")
    return a / b

help(dividir)  # En Colab te muestra el docstring

Help on function dividir in module __main__:

dividir(a, b)
    Devuelve a/b. Lanza ValueError si b == 0.



-help(objeto) es una *función incorporada de Python*.

-Sirve para mostrar *información de ayuda* sobre cualquier objeto de Python: funciones, clases, módulos, etc.

-Lo que hace es mostrar la *documentación interna (docstring)* del objeto, si existe.

### Pythonic
En la comunidad Python se habla de Pythonic code para referirse a código que:

-Aprovecha las *características únicas de Python* (listas por comprensión, manejo de excepciones, funciones integradas).

-Es *claro y legible* para otros programadores Python (legibilidad > trucos).

-Sigue la filosofía del lenguaje, resumida en el famoso *Zen of Python*.

Veamos algunos ejemplos.

### ¿Cuando usarlo?

-Cuando la sintaxis mejora la claridad sin sacrificar entendimiento.

Ejemplo: una lista por comprensión con una condición simple.

In [6]:
# Legible
pares = [x for x in range(10) if x % 2 == 0]

### ¿Cuando no usarlo?
Cuando la comprensión se vuelve demasiado compleja o difícil de seguir.

In [ ]:
# Esto es Pythonic pero ilegible
resultado = [f"{a}:{b}" for a in lista1 for b in lista2 if a > b and len(b) > 3]

# Mejor legibilidad
resultado = []
for a in lista1:
    for b in lista2:
        if a > b and len(b) > 3:
            resultado.append(f"{a}:{b}")

## Ejemplos adicionales
### Listas por comprensión

In [8]:
# No Pythonic
cuadrados = []
for i in range(10):
    cuadrados.append(i**2)

# Pythonic
cuadrados = [i**2 for i in range(10)]

### Iteración con enumerate

In [9]:
nombres = ["Ana", "Luis", "Sofía"]

# No Pythonic
i = 0
for nombre in nombres:
    print(i, nombre)
    i += 1

# Pythonic
for i, nombre in enumerate(nombres):
    print(i, nombre)

0 Ana
1 Luis
2 Sofía
0 Ana
1 Luis
2 Sofía


### Desempaquetado de variables

In [10]:
# Pythonic
nombre, edad, pais = ("Ana", 30, "México")

print(nombre)  # Ana
print(edad)    # 30
print(pais)    # México

Ana
30
México


Esto evita hacer:

In [11]:
datos = ("Ana", 30, "México")
nombre = datos[0]
edad = datos[1]
pais = datos[2]

También el uso de ignorar valores con _. El guión bajo indica "no me interesa este valor".

In [12]:
# Tenemos 4 valores pero solo nos importan 2
_, apellido, _, edad = ("Ana", "García", "CDMX", 30)

print(apellido)  # García
print(edad)      # 30

García
30


Otro uso es el de capturar " el resto" con *.

In [13]:
numeros = [1, 2, 3, 4, 5]

primero, *centrales, ultimo = numeros

print(primero)    # 1
print(centrales)  # [2, 3, 4]
print(ultimo)     # 5

1
[2, 3, 4]
5


### Uso de any y all

In [14]:
edades = [20, 18, 25]

# No Pythonic
mayores = True
for e in edades:
    if e < 18:
        mayores = False
        break

# Pythonic
mayores = all(e >= 18 for e in edades)

### Manejo de errores: EAFP vs LBYL

Esto, visto en un contenido anterior, forma parte de Pythonic también. En Python es común intentar y luego manejar excepciones.

Como en todo, dependerá el caso de uso, pero en la guía se te propone intentar EAFP como primer recurso.

In [15]:
persona = {"nombre": "Ana"}

# LBYL (mirar antes de saltar)
if "edad" in persona:
    edad = persona["edad"]
else:
    edad = "desconocida"

# EAFP (intentar y atrapar) - Esta es una mejor opción
try:
    edad = persona["edad"]
except KeyError:
    edad = "desconocida"

### ¿Por qué es importante saber Pythonic o conocerlo con más profundidad?

1-Python permite muchas formas de resolver un problema. Pythonic code ayuda a elegir la más clara.

2-Hace que el código sea más corto, expresivo y fácil de leer.

3-Refuerza la idea de que programar no es solo “hacer que se ejecute”, sino comunicar soluciones.

4-Al entrar al mundo laboral, te encontrarás con revisiones de código, y escribir de forma Pythonic es uno valioso conocimiento.

### En conclusión
## PEP 8

-Es una guía de estilo oficial (un documento escrito).

-Define reglas formales sobre cómo escribir Python:

-Indentación (4 espacios).

-Nombres de variables y funciones (snake_case).

-Nombres de clases (CamelCase).

-Longitud de líneas.

-Espacios en operadores y comas.

-Organización de importaciones.

-Su meta: que todo el mundo escriba con un estilo uniforme.

## Pythonic

-No es un documento, es una filosofía de cómo escribir Python de manera natural y legible.

-Se basa en el Zen of Python. Principios como:

-Simple is better than complex.

-Readability counts.

-Incluye patrones idiomáticos que aprovechan la elegancia de Python:

-List comprehensions.

-enumerate, zip, any, all.

-Desempaquetado.

-EAFP en lugar de LBYL.

-Su meta: que el código sea expresivo, legible y fácil de mantener.

-----------

# Instrucciones Finales


Entramos a nuestra semana final y con ello, comenzamos con el proyecto final.

A continuación hablaremos de las instrucciones, desarrollo, opciones y cómo podremos terminar el bootcamp satisfactoriamente.

## Primera etapa: Proyecto guiado
En esta etapa introductoria, lo que haremos es construir un proyecto, el cual será guiado. Es decir lo construiremos de cero a fin con la intención de repasar todos/la mayoría de los conceptos vistos en el Bootcamp.

Será desarrollado en Google Colab como siempre lo has hecho.

A continuación los pasos.

## Proyecto guiado: Sistema de inventarios

-Introducción y configuración

-Persistencia de datos

-Agregado de productos

-Listado y búsqueda de productos

-Actualización y borrado de productos

-Características adicionales

-Formato de entrega

Una vez realizado el proyecto guiado, se enviará normalmente vía nuestro canal de comunicación en WhatsApp.

Tan pronto termines este proyecto guiado, envíalo.

## Segunda etapa: Proyecto final
En esta etapa, la idea es desarrollar un proyecto el cual harás por tu cuenta.

El tema está relacionado con construir un catálogo de libros o películas.

Este catálogo puede ser del género que gustes o de manera general, puedes realizar un catálogo amplio de múltiples géneros.

Es abierto. Sé lo más creativo posible y puedes extenderte tanto como gustes.

A continuación, te presento los mínimos que debe de cumplir:

### Funcionalidad base
-Agregar elementos. Se pueden añadir libros/películas con todos los campos requeridos.

-Listar elementos. Muestra el catálogo completo de forma clara y legible.

-Buscar elementos. Permite encontrar por titulo (exacto o parcial).

-Actualizar elementos. Se pueden modificar campos existentes de un libro/película.

-Eliminar elementos. Se pueden borrar entradas del catálogo.

## Persistencia de datos

-Guardar/cargar JSON. El catálogo se puede almacenar y recuperar de un archivo.

-Opcional CSV. Exportación/lectura en formato tabular, si se implementa correctamente.

## Calidad de código

-Nombres claros y consistentes.

-Docstrings y comentarios útiles.

-Manejo de errores correcto. Evitar caídas con try/except.

-Organización modular y legibilidad.

## Características adicionales

-Reportes: conteo de elementos, valor total del catálogo, filtro por género/año.

-Búsquedas más avanzadas (parciales, por director/autor).

-Ordenar resultados (por título, año o precio).

-Validaciones extra (p. ej., anio debe ser numérico, precio positivo).

-Característica adicional que gustarías agregar.

## Opción A. Google Colab
Este formato es el mismo que hemos trabajado siempre. La idea es que, al terminar de construir tu proyecto, tengas el link y apliques el botón de “Share” para que pueda entrar a verlo y comentarlo.

## Opción B. Python en local con FastAPI
Este formato es diferente y no lo hemos visto en el bootcamp. Sin embargo, es un reto que vale la pena si quieres adentrarte con Python de una manera más profesional.

En este escenario, deberás terminar el proyecto que está listado a continuación:



## Proyecto: Catálogo de películas con Python y FastAPI

Este proyecto es guiado. Es decir, solo tendrás que seguir las instrucciones de inicio a fin.

Habrá temas adicionales como:

-Configuración e instalación de Python

-Ambientes virtuales

-pip e instalación de librerías

-Desarrollo de puntos de acceso

-Swagger para documentación y entrega

-Despliegue en Vercel

Al terminar, compartirás la url de tu despliegue, el cual es explicado en el proyecto guiado.

## ¿Cómo entregar?
Ahora, hablando de los entregables, es importante que consideres los puntos a continuación para entregar via WhatsApp.

1-Dependiendo de que tecnología escojas, enviarás el link del Google Colab o URL de Vercel.

2-Un video con tu celular explicando tu proyecto y qué hace.

-La duración del video debe ser de 5 - 7 minutos.

-Graba solo con tu voz. Ahora, si quieres enviarlo con tu cara dando una introducción, está bien, pero no es obligatorio. Tu cámara apuntará al código y explicarás parte por parte qué hace.

-Al comenzar, di tu nombre e inmediatamente comienza a explicar el proyecto. Habla de las funcionalidades, puntos fuertes de tu proyecto y muestra los resultados de tu código. No tienes que explicar línea por línea, solo las funciones y por qué planteaste la solución de esa forma.

-En la parte final del video, habla de la lección más importante que aprendiste con Python.

-Al terminar tu video, lo enviarás por WhatsApp a nuestro número de contacto al cual has enviado anteriormente tus progresos.

-Opcional. En caso de que no quieras grabar con tu celular y quieras dar una mejor presentación, puedes usar Loom y enviarme el link del video. Es gratuito y puedes usar tu computadora directamente para grabarlo. Podrás compartir tu escritorio a través del programa. https://www.loom.com/

Este será tu cierre de bootcamp a nivel de entregable.

## ¿Cuánto tiempo tengo?
Al ser la última semana, puedes tomarte el tiempo que necesites.

De preferencia, no tardes más de dos meses, ya que puedes perder el ritmo de estudio que llevabas y podrías abandonar el último paso.

Si llegas a tener algún bloqueo, puedes escribirme y haré lo posible por darte una respuesta a tu problema.

## Etapa final: Evaluación
Al entregar tu proyecto guiado y tu proyecto final, los revisaré y enviaré tu feedback final.

Adicionalmente, verificaré que tengas todos tus colabs entregados de las semanas pasadas (1-8).

Es importante señalarte que puede tardar entre 15 - 30 días a partir de la llegada del mensaje. Sin embargo, será revisado oportunamente a través de un mensaje de texto y audio.

Al estar todo en orden, te enviaré un link para que agendes conmigo. Hablaremos finalmente de:

-los pasos para tu activación de “badge” en Linkedin,

-tu mapa de aprendizaje de una manera más conversacional,

-darte mi agradecimiento por haber completado el bootcamp.